exonum/src/explorer/mod.rs

```diff
@@ -53,58 +53,69 @@ pub struct TxInfo {
     }
 
     /// Returns information about the transaction identified by the hash.
     pub fn tx_info(&self, tx_hash: &Hash) -> Result<Option<TxInfo>, ApiError> {
-        let b = self.blockchain.clone();
-        let snapshot = b.snapshot();
-        let schema = Schema::new(&snapshot);
-        let tx = schema.transactions().get(tx_hash);
-        let res = match tx {
-            None => None,
-            Some(raw_tx) => {
-                let box_transaction = self.blockchain.tx_from_raw(raw_tx.clone()).ok_or_else(|| {
-                    ApiError::InternalError(
-                        format!("Service not found for tx: {:?}", raw_tx).into(),
-                    )
-                })?;
-                let content = box_transaction.serialize_field().map_err(
-                    ApiError::InternalError,
-                )?;
-
-                let location = schema.tx_location_by_tx_hash().get(tx_hash).expect(
-                    &format!(
-                        "Not found tx_hash location: {:?}",
-                        tx_hash
-                    ),
-                );
-
-                let block_height = location.block_height();
-                let tx_index = location.position_in_block();
-                let proof = schema.block_txs(block_height).get_proof(tx_index);
-                let tx_info = TxInfo {
-                    content: content,
-                    location: location,
-                    proof_to_block_merkle_root: proof,
-                };
-                Some(tx_info)
-            }
-        };
-        Ok(res)
+        let schema = Schema::new(self.blockchain.snapshot());
+        Ok(if let Some(raw_tx) = schema.transactions().get(tx_hash) {
+            let box_transaction = self.blockchain.tx_from_raw(raw_tx.clone()).ok_or_else(|| {
+                ApiError::InternalError(
+                    format!("Service not found for tx: {:?}", raw_tx).into(),
+                )
+            })?;
+
+            let content = box_transaction.serialize_field().map_err(
+                ApiError::InternalError,
+            )?;
+
+            let location = schema.tx_location_by_tx_hash().get(tx_hash).expect(
+                &format!(
+                    "Not found tx_hash location: {:?}",
+                    tx_hash
+                ),
+            );
+
+            let proof_to_block_merkle_root = schema.block_txs(location.block_height())
+                .get_proof(location.position_in_block());
+
+            // Unwrap is OK here, because we already know that transaction is committed.
+            let status = match schema.transaction_results().get(tx_hash).unwrap() {
+                Ok(()) => TxStatus::Success,
+                Err(e) => TxStatus::Error(format!("{}", e))
+            };
+
+            Some(TxInfo {
+                content,
+                location,
+                proof_to_block_merkle_root,
+                status,
+            })
+        } else {
+            None
+        })
```

Comment:
This expression is overly nested to my taste :-) As an alternative, I would suggest

```Rust
let raw_tx = match schema.transactions().get(tx_hash) {
  Some(raw_tx) => raw_tx,
  None => return Ok(None),
};
// code code code

Ok(Some(TxInfo { ... }))
```